### PR TITLE
Fixes for a few XAtlas bugs (Tilers #508)

### PR DESCRIPTION
**Pull request by benedetto73**
Long term I suggest rearchitecting this part of XAtlas.

### First issue
In case of TaskGroups full (`TaskScheduler::m_groups`), the asserts were only in debug builds. The code would end up using some random memory for the group...

### Second issue
Waking up worker threads was broken.
When triggering the condition variable, if the thread was sleeping it was correctly woken up, but if it was running, at the end of the loop the code did not check the wakeup flag.

### Third issue
The shutdown flag was not checked in the inner worker thread loop.

### Fourth issue
No `try-catch` around worker threads loop.

### CODE REVIEW

File: xatlas.cpp [hunk 1]
```diff
@@ -3168,107 +3251,98 @@ class TaskScheduler
...
-			group.userData = userData;
-			group.ref = 0;
-			TaskGroupHandle handle;
-			handle.value = i;
-			return handle;
-		}
-		XA_DEBUG_ASSERT(false);
-		TaskGroupHandle handle;
-		handle.value = UINT32_MAX;
+		TaskGroup * group = new TaskGroup;
```
Comment (benedetto73) [inline xatlas.cpp]: Can use emplace into the syncmap?

File: xatlas.cpp [hunk 2]
```diff
@@ -70,13 +81,20 @@ Copyright (c) 2012 Brandon Pelfrey
...
-#define XA_ASSERT(exp) if (!(exp)) { XA_PRINT_WARNING("\rASSERT: %s %s %d\n", XA_XSTR(exp), __FILE__, __LINE__); }
+#define XA_ASSERT(exp) if (!(exp)) { XA_PRINT_WARNING("\r%s: ASSERT: %s %s %d\n", internal::getTID().c_str(), XA_XSTR(exp), __FILE__, __LINE__); }
 #endif
 
 #ifndef XA_DEBUG_ASSERT
 #define XA_DEBUG_ASSERT(exp) assert(exp)
 #endif
 
+#ifndef XA_ABORT
+#define XA_ABORT() std::fflush(stdout); abort();
```
Comment (benedetto73) [inline xatlas.cpp]: Consider also flushing stderr.

File: xatlas.cpp [hunk 3]
```diff
@@ -3185,38 +3189,45 @@ class sync_unordered_map {
...
+	// Returns a reference to its value if found.
+	// Throws exception if not found.
+	//
+	V& get (const K& k) {
 		std::lock_guard<std::mutex> guard(mx_mutex);
 
 		auto iter = m_data.find(k);
 		if (iter == m_data.end()) {
-			return std::nullopt;
+			throw std::runtime_error("Xatlas: TasHandle not found in syncmap.");
```
Comment (ptrgags) [inline xatlas.cpp]:
```suggestion
			throw std::runtime_error("Xatlas: TaskHandle not found in syncmap.");
```

**Review by ptrgags — APPROVED** [review]
@benedetto73 looks good, and I've run this in a couple cases, it doesn't hang like it used to!

I noticed one small typo in a comment, but I'll fix that through the UI and merge

**Pull request merged**